src/inset_state/densify.cpp

```diff
@@ -3,36 +3,84 @@
+//      coef_x * x + coef_y * y + coef_const = 0.
+// The function returns the unique intersection point between them. If this
+// intersection point does not exist, the function returns the point (-1, -1),
+// which is always outside of any graticule grid cell.
+Point calc_intersection(
+    const Point a,
+    const Point b,
+    const double coef_x,
+    const double coef_y,
+    const double coef_const
+) {
+  const auto result = CGAL::intersection(
+    Line(coef_x, coef_y, coef_const),
+    Segment(a, b)
+  );
   if (result) {
-    if (const Point* p = boost::get<Point>(&*result)) {
-      return (*p);
-    }
+    const Point* p = boost::get<Point>(&*result);
+    if (p) return (*p);
   }
+  return Point(-1.0, -1.0);
+}
 
-  // Value outside grid cell indicating that no intersection was found
-  return Point(-1, -1);
+// This function takes points `a` and `b` as arguments, which define a
+// segment. The function also takes the following arguments, which define the
+// types of diagonals for which we will calculate intersections:
+// - slope: the slope of every diagonal.
+// - base_intercept: the intercept that is the closest to 0 of a diagonal
+//   on the grid. This value is either 0, 0.25, or 0.5.
+// - step: what we need to add to each diagonal's intercept to obtain the
+//   next diagonal.
+void add_diag_inter(
+    std::set<Point, decltype(point_lesser)*> *intersections,
+    const Point a,
+    const Point b,
+    double slope,
+    double base_intercept,
+    double step,
+    const unsigned int lx,
+    const unsigned int ly
+) {
+  double intercept_start =
+    floor(std::min(a.y() - slope * a.x(), b.y() - slope * b.x()))
+    + base_intercept;
+  double intercept_end =
+    std::max(a.y() - slope * a.x(), b.y() - slope * b.x());
+  for (double d = intercept_start; d <= intercept_end; d += step) {
+
+    // We consider four types of diagonals:
+    // - 'normal': y = x + d,
+    // - 'antinormal': y = -x + d,
+    // - 'steep': y = 2x + d + base_intercept
+    //      (where base_intercept = 0.5),
+    // - 'antisteep': y = -2x + d + base_intercept
+    //      (where base_intercept = 0.5),
+    // - 'gentle': y = 0.5x + d + base_intercept,
+    //      (where base_intercept = 0.25),
+    // - 'antigentle': y = -0.5x + d + base_intercept,
+    //      (where base_intercept = 0.25),
+    // where d is the double increasing by `step` after each iteration of the
+    // loop.
+    // Steep and antisteep diagonals appear in graticule cells near x = 0
+    // and x = lx. Gentle and antigentle diagonals appear in graticules near
+    // y = 0 and y = ly.
+    Point inter = calc_intersection(a, b, slope, -1.0, d);
+    if (inter != Point(-1, -1) &&
+        ((abs(slope) == 2 && (inter.x() < 0.5 || inter.x() > (lx - 0.5))) ||
+         (abs(slope) == 0.5 && (inter.y() < 0.5 || inter.y() > (ly - 0.5))) ||
+         (abs(slope) == 1 && inter.x() >= 0.5 && inter.x() <= (lx - 0.5) &&
```

Comment:
I didn't include those intersections since the corner cells will never be concave, if I'm not mistaken. For example, the bottom-left corner cell will only have its top-right vertex moved by the integration. The bottom-left vertex will not move at all, and the top-left and bottom-right vertices will only mirror the vertical and horizontal movement of the top-right vertex respectively.

Reply:
I thought that since the corner cells remain convex, it shouldn't matter whether we add the diagonals or not? But most likely my logic is off; I've modified the `if` condition to add the `normal` and `antinormal` diagonal intersections in the corner cells. As a bonus, the `if` condition itself should be more readable in general now.